spring-integration-core/src/main/java/org/springframework/integration/history/MessageHistory.java

```diff
@@ -73,7 +78,26 @@ public static MessageHistory read(Message<?> message) {
 					new ArrayList<Properties>(previousHistory) : new ArrayList<Properties>();
 			components.add(metadata);
 			MessageHistory history = new MessageHistory(components);
-			message = messageBuilderFactory.fromMessage(message).setHeader(HEADER_NAME, history).build();
+
+			if (message instanceof MutableMessage) {
+				message.getHeaders().put(HEADER_NAME, history);
+			}
+			else if (message instanceof ErrorMessage) {
+				IntegrationMessageHeaderAccessor headerAccessor = new IntegrationMessageHeaderAccessor(message);
+				headerAccessor.setHeader(HEADER_NAME, history);
+				Throwable payload = ((ErrorMessage) message).getPayload();
+				ErrorMessage errorMessage = new ErrorMessage(payload, headerAccessor.toMessageHeaders());
+				message = (Message<T>) errorMessage;
+			}
+			else if (message instanceof AdviceMessage) {
+				IntegrationMessageHeaderAccessor headerAccessor = new IntegrationMessageHeaderAccessor(message);
+				headerAccessor.setHeader(HEADER_NAME, history);
+				message = new AdviceMessage<T>(message.getPayload(), headerAccessor.toMessageHeaders(),
+						((AdviceMessage) message).getInputMessage());
+			}
+			else {
+				message = messageBuilderFactory.fromMessage(message).setHeader(HEADER_NAME, history).build();
```

Comment:
Perhaps add a DEBUG log (or WARN?) for not-recognized message types (e.g. User) ?